File: trimmed_match/estimator_test.py
# Copyright 2020 Google LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================

"""Tests for the python CLIF wrapper for trimmedmatch."""

from absl.testing import absltest
from absl.testing import parameterized

from trimmed_match import estimator


def _ConstructReport(**kwargs):
  """Creates a Report."""
  if kwargs.keys() != set([
      "estimate", "std_error", "trim_rate", "confidence", "conf_interval_low",
      "conf_interval_up", "epsilons", "trimmed_pairs_indices",
      "candidate_results"
  ]):
    raise KeyError("The arguments are incorrect")

  temp_results = []
  for x in kwargs["candidate_results"]:
    temp = estimator.TrimAndError()
    temp.trim_rate = x[0]
    temp.iroas = x[1]
    temp.std_error = x[2]
    temp_results.append(temp)

  return estimator.Report(
      kwargs.get("estimate"), kwargs.get("std_error"), kwargs.get("trim_rate"),
      kwargs.get("confidence"), kwargs.get("conf_interval_low"),
      kwargs.get("conf_interval_up"), kwargs.get("epsilons"),
      kwargs.get("trimmed_pairs_indices"), temp_results)


class EstimatorTest(absltest.TestCase):

  def setUp(self):
    """Setup reusable data."""
    super(EstimatorTest, self).setUp()

    # test example
    self._delta_response = [1.0, 10.0, 3.0, 8.0]
    self._delta_cost = [1.0, 5.0, 2.0, 5.0]
    self._iroas0 = sum(self._delta_response) / sum(self._delta_cost)
    self._iroas1 = sum(self._delta_response[2:]) / sum(self._delta_cost[2:])

    self._report_no_trim = _ConstructReport(
        estimate=self._iroas0,
        std_error=0.138,
        trim_rate=0.0,
        confidence=0.9,
        conf_interval_low=1.250,
        conf_interval_up=1.905,
        epsilons=[
            self._delta_response[i] - self._delta_cost[i] * self._iroas0
            for i in range(0, len(self._delta_response))
        ],
        trimmed_pairs_indices=set(),
        candidate_results=[[0, 1.692, 0.138]])

    self._report_trim1 = _ConstructReport(
        estimate=self._iroas1,
        std_error=0.041,
        trim_rate=0.20,
        confidence=0.9,
        conf_interval_low=-29.374,
        conf_interval_up=1.619,
        epsilons=[
            self._delta_response[i] - self._delta_cost[i] * self._iroas1
            for i in range(0, len(self._delta_response))
        ],
        trimmed_pairs_indices=set([0, 1]),
        candidate_results=[[0.20, 1.571, 0.041]])

    self._report_auto_trim = _ConstructReport(
        estimate=self._iroas1,
        std_error=0.041,
        trim_rate=0.25,
        confidence=0.9,
        conf_interval_low=-29.374,
        conf_interval_up=1.619,
        epsilons=[
            self._delta_response[i] - self._delta_cost[i] * self._iroas1
            for i in range(0, len(self._delta_response))
        ],
        trimmed_pairs_indices=set([0, 1]),
        candidate_results=[[0.0, 1.692, 0.138], [0.25, 1.571, 0.041]])

  def AssertReportEqual(self, expected, report):
    """Compares Report with expected values."""
    for attr in [
        "estimate", "std_error", "trim_rate", "confidence", "conf_interval_low",
        "conf_interval_up"
    ]:
      self.assertAlmostEqual(
          getattr(expected, attr), getattr(report, attr), places=3)
    self.assertEqual(
        len(expected.candidate_results), len(report.candidate_results))
    self.assertEqual(len(expected.epsilons), len(report.epsilons))
    for e1, e2 in zip(expected.epsilons, report.epsilons):
      self.assertAlmostEqual(e1, e2, places=3)
    self.assertSetEqual(expected.trimmed_pairs_indices,
                        report.trimmed_pairs_indices)
    for i in range(0, len(expected.candidate_results)):
      for attr in ["trim_rate", "iroas", "std_error"]:
        self.assertAlmostEqual(
            getattr(expected.candidate_results[i], attr),
            getattr(report.candidate_results[i], attr),
            places=3)

  def testTrimmedMatchValueError(self):
    # if max_trim_rate is negative
    with self.assertRaisesRegex(ValueError, "max_trim_rate is negative."):
      _ = estimator.TrimmedMatch(self._delta_response, self._delta_cost, -0.1)
    # if delta_response and delta_delta have different lengths
    with self.assertRaisesRegex(
        ValueError, "Lengths of delta_response and delta_spend differ."):
      _ = estimator.TrimmedMatch(self._delta_response, self._delta_cost + [1.0])
    # if confidence is outside of (0, 1]
    tm = estimator.TrimmedMatch(self._delta_response, self._delta_cost)
    with self.assertRaisesRegex(ValueError,
                                r"Confidence is outside of \(0, 1\]"):
      _ = tm.Report(-0.5, 0.0)

  def testCalculateEpsilons(self):
    """Tests _CalculateEpsilons."""
    tm = estimator.TrimmedMatch(self._delta_response, self._delta_cost, 0.25)
    report = tm.Report(0.90, 0.0)
    expected = [
        self._delta_response[i] - self._iroas0 * self._delta_cost[i]
        for i in range(0, len(self._delta_response))
    ]
    self.assertEqual(len(expected), len(report.epsilons))
    for i in range(0, len(expected)):
      self.assertAlmostEqual(expected[i], report.epsilons[i], places=3)

  def testReporValueError(self):
    tm = estimator.TrimmedMatch(self._delta_response, self._delta_cost, 0.25)

    @parameterized.parameters((-0.1, 0.1), (1.1, 0.1), (0.8, 0.5))
    def _(self, confidence, trim_rate):
      with self.assertRaises(ValueError):
        tm.Report(confidence, trim_rate)

  def testTrimmedMatchCase(self):
    """Tests with various trim rates."""
    tm = estimator.TrimmedMatch(self._delta_response, self._delta_cost, 0.25)

    @parameterized.parameters((self._report_no_trim, 0.0),
                              (self._report_trim1, 0.20),
                              (self._report_auto_trim, -1))
    def _(self, expected, trim_rate):
      self.AssertReportEqual(expected, tm.Report(0.90, trim_rate))

  def testTrimmedMatchTiedSpend(self):
    tm = estimator.TrimmedMatch([1, 2, 3, 4, 5], [1, 1, 1, 1, 1])
    self.assertAlmostEqual(tm.Report().estimate, 3.0)

  def testTrimmedMatchTiedThetas(self):
    tm = estimator.TrimmedMatch([1, 2, 3, 4, 5], [1, 2, 3, 4, 5])
    self.assertAlmostEqual(tm.Report().estimate, 1.0)

  def testTrimmedMatchTiedThetasConstant(self):
    tm = estimator.TrimmedMatch([1, 1, 1, 1, 1], [1, 1, 1, 1, 1])
    self.assertAlmostEqual(tm.Report().estimate, 1.0)

  def testTrimmedMatchZeroSpend(self):
    with self.assertRaisesRegex(ValueError,
                                "delta_spends are all too close to 0!"):
      _ = estimator.TrimmedMatch([1, 2, 3, 4, 5], [0, 0, 0, 0, 0])


if __name__ == "__main__":
  absltest.main()

File: trimmed_match/core/estimator_test.cc
/*
* Copyright 2020 Google LLC.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     https://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
* ============================================================================
*/
#include "trimmed_match/core/estimator.h"

#include <cmath>

#include "gtest/gtest.h"

namespace trimmedmatch {
namespace {

void CompareReports(const Result& expected,
                    const absl::StatusOr<Result>& result,
                    const double epsilon = 1e-6) {
  if (result.ok()) {
    EXPECT_NEAR(expected.estimate, result->estimate, epsilon);
    EXPECT_NEAR(expected.std_error, result->std_error, epsilon);
    EXPECT_NEAR(expected.trim_rate, result->trim_rate, epsilon);
    EXPECT_NEAR(expected.conf_interval_low, result->conf_interval_low, epsilon);
    EXPECT_NEAR(expected.conf_interval_up, result->conf_interval_up, epsilon);
    EXPECT_EQ(expected.candidate_results.size(),
              result->candidate_results.size());

    for (size_t i = 0; i < result->candidate_results.size(); ++i) {
      EXPECT_NEAR(expected.candidate_results[i].iroas,
                  result->candidate_results[i].iroas, epsilon);
    }
  }
}

class EstimatorInternalTest : public ::testing::Test {
 protected:
  EstimatorInternalTest()
      : delta_response1_({1.0, 10.0, 3.0, 8.0}),
        delta_cost1_({1.0, 5.0, 2.0, 5.0}),
        trimmed_match1_(delta_response1_, delta_cost1_),
        trimmed_match2_(
            {552,   -710,  -961,   -1218,  -1293,  -1824, 1816,   2073,
             -2416, -2559, 2603,   3073,   3288,   3137,  4016,   3905,
             4032,  -4649, 5397,   -6054,  6451,   7735,  8507,   -9186,
             9700,  9987,  -11255, -13828, -14950, 20957, -38666, 43407},
            {89143,    -111084, -163240,  -201668, -213271, -311807,  313531,
             350041,   -425882, -439139,  431609,  509310,  554237,   548335,
             677060,   654552,  686768,   -812570, 940149,  -1019919, 1086827,
             1333913,  1458086, -1550027, 1676807, 1692128, -1927924, -2341819,
             -2533673, 3570161, -6614810, 7440206},
            0.30) {}

  const std::vector<double> delta_response1_;
  const std::vector<double> delta_cost1_;
  const TrimmedMatch trimmed_match1_;
  const TrimmedMatch trimmed_match2_;
};

TEST(TrimmedMatchInitialization, TrimmedMatchInvalidInput) {
  EXPECT_DEATH(auto result = TrimmedMatch({1, 2, 3}, {1, 2}), "");
}

TEST_F(EstimatorInternalTest, CalculateIroasEmptyRoot) {
  TrimmedMatch trimmed_match({1, 2, 3, 4}, {-10, -1, 1, 10});
  EXPECT_FALSE(trimmed_match.CalculateIroas(0.1).ok());
  EXPECT_EQ(
      trimmed_match.CalculateIroas(0.1).status(),
      absl::InternalError(
          "We could not find a root for the TM equation. One likely reason is "
          "that the incremental cost for the untrimmed geo pairs is 0."));
}

TEST_F(EstimatorInternalTest, InvalidInput) {
  EXPECT_FALSE(trimmed_match1_.CalculateIroas(-0.25).ok());
  EXPECT_EQ(trimmed_match1_.CalculateIroas(-0.25).status(),
            absl::InvalidArgumentError(
                "Trim rate must be in (0,0.25), but got -0.25"));
}

TEST_F(EstimatorInternalTest, CalculateIroasNoTrim) {
  double total_delta_response = 0.0, total_delta_cost = 0.0;

  for (size_t i = 0; i < delta_response1_.size(); ++i) {
    total_delta_response += delta_response1_[i];
    total_delta_cost += delta_cost1_[i];
  }

  EXPECT_NEAR(total_delta_response / total_delta_cost,
              *trimmed_match1_.CalculateIroas(0.0), 1e-6);
}

TEST_F(EstimatorInternalTest, CalculateIroasWithTrim) {
  double total_delta_response = 0.0, total_delta_cost = 0.0;

  // First two pairs are trimmed.
  for (size_t i = 2; i < delta_response1_.size(); ++i) {
    total_delta_response += delta_response1_[i];
    total_delta_cost += delta_cost1_[i];
  }

  EXPECT_NEAR(total_delta_response / total_delta_cost,
              *trimmed_match1_.CalculateIroas(0.25), 1e-6);
}

TEST_F(EstimatorInternalTest, CalculateStandardErrorNoTrim) {
  const double iroas = 0.0;
  double sum_square_delta_response = 0.0;
  double sum_delta_cost = 0.0;
  for (size_t i = 0; i < delta_response1_.size(); ++i) {
    sum_square_delta_response += Square(delta_response1_[i]);
    sum_delta_cost += delta_cost1_[i];
  }

  EXPECT_NEAR(std::sqrt(sum_square_delta_response) / sum_delta_cost,
              trimmed_match1_.CalculateStandardError(0.0, iroas), 1e-6);
}

TEST_F(EstimatorInternalTest, CalculateStandardErrorWithTrim) {
  const double iroas = 1.0;
  double sum_square_delta_response = 0.0;
  double sum_delta_cost = 0.0;

  // First two pairs are trimmed.
  for (size_t i = 2; i < delta_response1_.size(); ++i) {
    sum_square_delta_response +=
        Square(delta_response1_[i] - iroas * delta_cost1_[i]) * 2;
    sum_delta_cost += delta_cost1_[i];
  }

  EXPECT_NEAR(std::sqrt(sum_square_delta_response) / sum_delta_cost,
              trimmed_match1_.CalculateStandardError(0.25, iroas), 1e-6);
}

TEST_F(EstimatorInternalTest, TrimmedMatch1NoTrim) {
  const Result expected = {
      1.692, 0.138, 0, 0.9, 1.250, 1.905, {{0, 1.692, 0.138}}};
  const auto result = trimmed_match1_.Report(1.64485363, 0.0);

  CompareReports(expected, result, 0.001);
}

TEST_F(EstimatorInternalTest, TrimmedMatch1WithTrim) {
  const Result expected = {
      1.571, 0.041, 0.25, 0.9, -29.375, 1.619, {{0.25, 1.571, 0.041}}};
  const auto result = trimmed_match1_.Report(1.64485363, 0.25);

  CompareReports(expected, result, 0.001);
}

TEST_F(EstimatorInternalTest, TrimmedMatch2NoTrim) {
  const Result expected = {0.005811,
                           0.000117,
                           0,
                           0.9,
                           0.005233,
                           0.006389,
                           {{0, 0.005811, 0.000117}}};
  const auto result = trimmed_match2_.Report(1.64485363, 0.0);

  CompareReports(expected, result);
}

TEST_F(EstimatorInternalTest, TrimmedMatch2WithTrim) {
  const Result expected = {0.005815,
                           0.000134,
                           0.03125,
                           0.9,
                           0.005155,
                           0.006475,
                           {{0.03125, 0.005815, 0.000134}}};
  const auto result = trimmed_match2_.Report(1.64485363, 1.0 / 32);

  CompareReports(expected, result);
}

TEST_F(EstimatorInternalTest, TrimmedMatch2WithOptimTrim) {
  const std::vector<TrimAndError> candidates = {
      {0, 0.00581123, 0.000117108},   {0.03125, 0.005815, 0.000133689},
      {0.0625, 0.00582772, 0.000094}, {0.09375, 0.00583428, 0.000073},
      {0.125, 0.00583461, 0.000062},  {0.15625, 0.00583399, 0.000061},
      {0.1875, 0.00583305, 0.000030}, {0.21875, 0.00583258, 0.000031},
      {0.25, 0.005832, 0.000029},     {0.28125, 0.00585823, 0.000069}};

  const Result expected = {0.005833, 0.00003,  0.1875,    0.9,
                           0.005684, 0.005982, candidates};
  const auto result = trimmed_match2_.Report(1.64485363, -1);

  CompareReports(expected, result);
}

}  // namespace
}  // namespace trimmedmatch
